ebpf/state.go

```diff
@@ -98,134 +114,125 @@ func (ns *networkState) getClients() []string {
 	if err != nil {
-		log.Errorf("%s", err)
+		log.Warn("failed to create byte key: %s", err)
 		return
 	}
-	key := string(rawKey)
 
-	for id := range ns.clients {
-		if prev, ok := ns.clients[id].closedConnections[key]; ok {
+	for _, client := range ns.clients {
+		if prev, ok := client.closedConnections[string(key)]; ok {
 			// Entry is already here, add the old connection to override connections
 			// For later aggregation
-			newOverride := statsFromConn(*prev)
-			if override, ok := ns.clients[id].overrideConnections[key]; ok {
+			newOverride := statsFromConn(*prev, now)
+			if override, ok := client.overrideConnections[string(key)]; ok {
 				// If we already have an override aggregate the two overrides
-				newOverride = aggregateStats(override, newOverride)
+				newOverride = aggregateStats(override, newOverride, now)
 			}
-			ns.clients[id].overrideConnections[key] = newOverride
+			client.overrideConnections[string(key)] = newOverride
 		}
 
 		// We only store the pointer to the connection, when it will be cleared for each client it will get GCed
-		ns.clients[id].closedConnections[key] = &conn
+		client.closedConnections[string(key)] = &conn
 	}
 }
 
 // closedConns returns the closed connections for the given client and takes care of updating last fetch
 // the provided client is supposed to exist
 func (ns *networkState) closedConns(clientID string) []ConnectionStats {
-	conns := []ConnectionStats{}
+	client := ns.clients[clientID]
+	conns := make([]ConnectionStats, 0, len(client.closedConnections))
 
-	ns.Lock()
-	defer ns.Unlock()
-
-	for key, conn := range ns.clients[clientID].closedConnections {
-		// We dereference to avoid modifying the underlying connection
-		// since it's shared with other clients
+	for key, conn := range client.closedConnections {
 		c := *conn
 
 		// First check if we have an override stored
 		// If we do aggregate it and delete the override
-		if override, ok := ns.clients[clientID].overrideConnections[key]; ok {
-			aggregateConnAndStat(&c, override)
-			delete(ns.clients[clientID].overrideConnections, key)
+		if override, ok := client.overrideConnections[key]; ok {
+			aggregateConnAndStat(&c, &override)
+			delete(client.overrideConnections, key)
 		}
 
-		// Total defaults to 0 if it's not stored
-		prev := sentRecvStats{}
-		if _, ok := ns.clients[clientID].stats[key]; ok {
-			prev = *ns.clients[clientID].stats[key]
-			delete(ns.clients[clientID].stats, key)
+		if stats, ok := client.stats[key]; ok {
+			c.LastSentBytes = c.MonotonicSentBytes - stats.totalSent
+			c.LastRecvBytes = c.MonotonicRecvBytes - stats.totalRecv
+			c.LastRetransmits = c.MonotonicRetransmits - stats.totalRetransmits
+			delete(client.stats, key)
+		} else {
+			c.LastSentBytes = c.MonotonicSentBytes
+			c.LastRecvBytes = c.MonotonicRecvBytes
+			c.LastRetransmits = c.MonotonicRetransmits
 		}
 
-		// Update last stats
-		c.LastSentBytes = prev.lastSent + c.MonotonicSentBytes - prev.totalSent
```

Comment:
This bit was actually a bug that passed tests due to unrealistic usage of the state interface. All fixed now!